File: companies-leetcode/amazon/linked-lists/reverse-node-in-k-group.js
/*
Given a linked list, reverse the nodes of a linked list k at a time and return its modified list.

k is a positive integer and is less than or equal to the length of the linked list. If the number of nodes is not a multiple of k then left-out nodes in the end should remain as it is.

Example:

Given this linked list: 1->2->3->4->5

For k = 2, you should return: 2->1->4->3->5

For k = 3, you should return: 3->2->1->4->5

Note:

Only constant extra memory is allowed.
You may not alter the values in the list's nodes, only nodes itself may be changed.
*/

// 75ms. 70th percentile
// Uses k memory...not sure if that's ok...
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
        let res = new ListNode(-1);
        let builder = res;
        
        while (head !== null) {
            let currentKGroup = [];
            
            for (let i = 0; i < k; i++) {
                currentKGroup.push(head.val);
                head = head.next;

                if (head === null && i < k-1) {
                    
                    for (let val of currentKGroup) {
                        builder.next = new ListNode(val);
                        builder = builder.next;
                    }
                    return res.next;
                } 
            }
            
            currentKGroup.reverse();
            for (let val of currentKGroup) {
                builder.next = new ListNode(val);
                builder = builder.next;
            }
        }
        
        return res.next;
    }
        
// attempt 2: 
/*
The above answer isn't valid. We aren't supposed to make new nodes. Above I was thinking of reversing as "switch the first
and the last nodes. Then the second and second to last...and so on". The key to this is to think of it simply as "makes each
node point to the one before it". Do that k times. Then attach that subchain to the result of reversing the next subchain
and so on. 

Just tricky handling all the pointers.
*/
// 72ms. 82nd percentile
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    if (k < 2 || head === null) {
        return head;
    }

    let reverseKGroup = function(nodeBeforeGroup, nodeAfterGroup) {
        let previous = nodeBeforeGroup;
        let current = nodeBeforeGroup.next;
        let next;
        let newEndOfGroup = current;
        while (current !== nodeAfterGroup) {
            next = current.next;
            current.next = previous;
            previous = current;
            current = next;
        }
        
        nodeBeforeGroup.next = previous;
        newEndOfGroup.next = nodeAfterGroup;
        return newEndOfGroup;
    }
    
    let dummy = new ListNode(-1);
    dummy.next = head;
    let nodeBeforeGroup = dummy;
    let i = 0;
    while (head !== null) {
        i += 1;
        
        if (i % k === 0) {
            nodeBeforeGroup = reverseKGroup(nodeBeforeGroup, head.next);
            head = nodeBeforeGroup.next;
        } else {
            head = head.next
        }
    }
    
    return dummy.next;
};

File: common-problems-leetcode/hard/reverse-node-in-k-group.py
"""
Given a linked list, reverse the nodes of a linked list k at a time and return its modified list.

k is a positive integer and is less than or equal to the length of the linked list. If the number of nodes is not a 
multiple of k then left-out nodes in the end should remain as it is.

Example:

Given this linked list: 1->2->3->4->5

For k = 2, you should return: 2->1->4->3->5

For k = 3, you should return: 3->2->1->4->5

Note:

Only constant extra memory is allowed.
You may not alter the values in the list's nodes, only nodes itself may be changed.
"""
# 52ms. 89 percentile.
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if k < 2 or not head:
            return head
        
        runner, length = head, 0
        while runner:
            length, runner = length + 1, runner.next
        
        res = current = ListNode(-1)
        res.next = head
        for _ in range(length // k):
            current = self._reverseKNodes(current, k)
        return res.next
    
    def _reverseKNodes(self, nodeBeforeGroup, k):
        previous, current, after = nodeBeforeGroup, nodeBeforeGroup.next, None
        newEndOfGroup = current
        while current and k != 0:
            after = current.next
            current.next = previous
            previous = current
            current = after
            k -= 1
        
        nodeBeforeGroup.next = previous
        newEndOfGroup.next = current
        return newEndOfGroup


"""
Notes:
When you reverse a linkedlist, you can initialize previous to None and make the first node point back to it. This is a little
bit different, because we need to keep track of the node before the group we're reversing so that we can make the end
of the reversed section point back to it.

Also, as far as I can tell, there is no way to avoid going over the list twice. You only want to reverse a section if
there are enough nodes to complete a full section. This implies that you have to look ahead k nodes befoe doing a reversal.
Or equivalently, count all the nodes at the beginning. 
"""